reactor-netty-http/src/test/java/reactor/netty/http/client/Http2PooledConnectionProviderCustomMetricsTest.java

```diff
@@ -0,0 +1,190 @@
+ */
+package reactor.netty.http.client;
+
+import io.netty.handler.ssl.SslContext;
+import io.netty.handler.ssl.SslContextBuilder;
+import io.netty.handler.ssl.util.InsecureTrustManagerFactory;
+import io.netty.handler.ssl.util.SelfSignedCertificate;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import reactor.core.publisher.Mono;
+import reactor.netty.DisposableServer;
+import reactor.netty.http.server.HttpServer;
+import reactor.netty.resources.ConnectionProvider;
+import reactor.util.annotation.Nullable;
+
+import javax.net.ssl.SSLException;
+import java.net.SocketAddress;
+import java.security.cert.CertificateException;
+import java.time.Duration;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.IntStream;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static reactor.netty.http.HttpProtocol.H2;
+
+class Http2PooledConnectionProviderCustomMetricsTest {
+
+	static SslContext sslServer;
+	static SslContext sslClient;
+
+	@BeforeEach
+	void setUp() throws CertificateException, SSLException {
+		SelfSignedCertificate ssc = new SelfSignedCertificate();
+		sslServer = SslContextBuilder.forServer(ssc.certificate(), ssc.privateKey())
+				.build();
+		sslClient = SslContextBuilder.forClient()
+				.trustManager(InsecureTrustManagerFactory.INSTANCE)
+				.build();
+	}
+
+	@Test
+	void measureActiveStreamsSize() throws InterruptedException {
+		AtomicBoolean isRegistered = new AtomicBoolean();
+		AtomicBoolean isDeregistered = new AtomicBoolean();
+		AtomicReference<Http2ConnectionPoolMetrics> metrics = new AtomicReference<>();
+
+		DisposableServer disposableServer = HttpServer.create()
+				.port(5678)
+				.protocol(H2)
+				.wiretap(true)
+				.secure(spec -> spec.sslContext(sslServer))
+				.handle((req, resp) -> resp.sendString(
+								Mono.delay(Duration.ofSeconds(10))
+										.then(Mono.just("test"))
+						)
+				)
+				.bindNow();
+
+		CustomHttp2MeterRegistrar registrar = new CustomHttp2MeterRegistrar(isRegistered, isDeregistered, metrics);
+		ConnectionProvider pool = ConnectionProvider.builder("custom-pool-2")
+				.metrics(true, () -> registrar)
+				.maxConnections(10)
+				.build();
+
+		CountDownLatch latch = new CountDownLatch(5);
+		HttpClient httpClient = HttpClient.create(pool);
+		IntStream.range(0, 5)
+				.forEach(unUsed -> httpClient.remoteAddress(disposableServer::address)
+						.protocol(H2)
+						.secure(spec -> spec.sslContext(sslClient))
+						.doOnConnected(connection -> latch.countDown())
+						.get()
+						.uri("/")
+						.responseSingle((resp, bytes) -> bytes.asString())
+						.subscribe()
+				);
+		latch.await();
+
+		assertThat(isRegistered.get()).isTrue();
+		assertThat(metrics.get().activeStreamSize()).isEqualTo(5);
+
+		pool.dispose();
+
+		assertThat(isDeregistered.get()).isTrue();
+
+		disposableServer.disposeNow();
```

Comment:
Let's dispose the pool and the server in a finally block.

Reply:
@violetagg 

I added `assertThat(isDeregistered.get()).isTrue()` to verify that `deRegisterMetrics()` is properly called when the pool is disposed, as below review. (https://github.com/reactor/reactor-netty/pull/3632#discussion_r1963812437)

I think that `pool.dispose();` should be kept **before** `assertThat(isDeregistered.get()).isTrue()` to verify `deRegisterMetrics()`.
what are your thoughts on this?

On the other hand, I will remove `disposableServer.disposeNow();` since it is already handled in BaseHttpTest's teardown."

Reply:
agree but we still need to dispose the pool and to check for deregistration in a finally block